plugins/signals-backend/src/service/router.ts

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { errorHandler } from '@backstage/backend-common';
+import express, { NextFunction, Request, Response } from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import { SignalService } from '@backstage/plugin-signals-node';
+import * as https from 'https';
+import http from 'http';
+
+/** @public */
+export interface RouterOptions {
+  logger: Logger;
+  service: SignalService;
+}
+
+/** @public */
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  const { logger, service } = options;
+  let subscribed = false;
+  const upgradeMiddleware = (req: Request, _: Response, next: NextFunction) => {
+    const server: https.Server | http.Server = (req.socket as any)?.server;
+    if (
+      !server ||
+      !req.headers ||
+      req.headers.upgrade === undefined ||
+      req.headers.upgrade.toLowerCase() !== 'websocket'
+    ) {
+      next();
+      return;
+    }
+
+    if (!subscribed) {
+      subscribed = true;
+      server.on('upgrade', async (request, socket, head) => {
```

Comment:
I am a bit worried if the KubernetesProxy will still work after this as it does the exact same thing for the server here https://github.com/chimurai/http-proxy-middleware/blob/master/src/http-proxy-middleware.ts#L90

And not sure if there is actually a way to bypass it.

Reply:
Actually I think it's ok as the HTTP proxy middleware also checks the path and does nothing if it doesn't match the configured one.